moodle/Sniffs/Commenting/PHPDocTypesSniff.php

```diff
@@ -0,0 +1,960 @@
+    protected ?object $commentpending = null;
+
+    /** @var int how long until we dispose of a pending comment */
+    protected int $commentpendingcounter = 0;
+
+    /** @var ?(\stdClass&object{tags: array<string, string[]>}) PHPDoc comment for current declaration */
+    protected ?object $comment = null;
+
+    /** @var array{'code': ?array-key, 'content': string, 'scope_opener'?: int, 'scope_closer'?: int}
+     * the current token */
+    protected array $token = ['code' => null, 'content' => ''];
+
+    /** @var array{'code': ?array-key, 'content': string, 'scope_opener'?: int, 'scope_closer'?: int}
+     * the previous token */
+    protected array $tokenprevious = ['code' => null, 'content' => ''];
+
+    /**
+     * Register for open tag (only process once per file).
+     * @return array-key[]
+     */
+    public function register(): array {
+        return [T_OPEN_TAG];
+    }
+
+    /**
+     * Processes PHP files and perform PHPDoc type checks with file.
+     * @param File $phpcsfile The file being scanned.
+     * @param int $stackptr The position in the stack.
+     * @return void
+     */
+    public function process(File $phpcsfile, $stackptr): void {
+
+        // Check we haven't already done this file.
+        if ($phpcsfile == $this->file) {
+            return;
+        }
+
+        try {
+            $this->file = $phpcsfile;
+            $this->tokens = $phpcsfile->getTokens();
+            $this->artifacts = [];
+
+            // Gather atifact info.
+            $this->pass = 1;
+            $this->typeparser = null;
+            $this->fileptr = $stackptr;
+            $this->processPass();
+
+            // Check the PHPDoc types.
+            $this->pass = 2;
+            $this->typeparser = new PHPDocTypeParser($this->artifacts);
+            $this->fileptr = $stackptr;
+            $this->processPass();
+        } catch (\Exception $e) {
+            $this->file->addError(
+                'The PHPDoc type sniff failed to parse the file.  PHPDoc type checks were not performed.',
+                $this->fileptr < count($this->tokens) ? $this->fileptr : $this->fileptr - 1,
+                'phpdoc_type_parse'
+            );
+        }
+    }
+
+    /**
+     * A pass over the file.
+     * @return void
+     * @phpstan-impure
+     */
+    protected function processPass(): void {
+        $this->scopes = [(object)['type' => 'root', 'namespace' => '', 'uses' => [], 'templates' => [],
+                        'classname' => null, 'parentname' => null, 'opened' => true, 'closer' => null]];
+        $this->tokenprevious = ['code' => null, 'content' => ''];
+        $this->fetchToken();
+        $this->commentpending = null;
+        $this->comment = null;
+
+        while ($this->token['code']) {
+            // Skip irrelevant tokens.
+            while (
+                !in_array(
+                    $this->token['code'],
```

Comment:
Check out the tokens in `\PHP_CodeSniffer\Util\Tokens` and https://phpcsutils.com/phpdoc/classes/PHPCSUtils-Tokens-Collections.html to see if there is a similar list. I'm not sure what this is intended to do.

Reply:
If you're talking about the scope_opener / scope_closer stuff, it's to deal with alternative syntax, where a colon can be a scope opener.  I had thought this could apply to classes and functions, but looking it up again, perhaps it's just control structures.

Reply:
Actually, perhaps it's good to look at the scope_openers/_closers anyway.  It looks like phpcs doesn't mark spurious close curly brackets at the end of a file as scope_closers.  If there are any other differences, then looking at the scope_openers/closers instead of token types may make the sniff better match up with the way phpcs parses a broken PHP file.

Reply:
Oh, and I'll look at those lists.

Reply:
We really don't want to write a parser within phpcs, which is itself a parser. We shoudl really aim to have one toolset here.

Reply:
And I was more referring to the list of inline T_ constants.

See the Tokens COllection and the Tokens in phpcs. These are largely already defined and _maintained_ there.

Reply:
I need to know what's nested in what to figure out which templates apply.  I guess phpcs parses the file and figures this out, but as far as I can tell, there isn't a good way to access this information as an abstract syntax tree or something.

Reply:
Oh, I also need to know the nesting to figure out if a USE token is for importing other PHP files, or class trait usage, or whatever it is that functions do with it.

Reply:
I've made use of lists from Tokens.